From 5d44f4cfe4c62e44e487385a35664219ac0f75d2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matthew=20=F0=9F=AA=B7?= <matthew@urgero.net>
Date: Sat, 14 Dec 2024 05:55:00 -0600
Subject: [PATCH] Winter Solstice 2024 Upgrade (Ruth epoch) (#5)

There are no changes that will activate with this upgrade.

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 6f488b931f90132720f2dbf1476c15faa1e3e846
Recorded-replay-commit: e2c81fc5e27a3407c0942d87940dcfd204f2fabc

diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 9b2d07c904..4290c3075c 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -112,3 +112,13 @@ bool IsJudgesEnabled(const Consensus::Params &params,
     return pindexPrev->GetMedianTimePast() >=
            gArgs.GetIntArg("-judgesactivationtime", params.judgesActivationTime);
 }
+
+bool IsRuthEnabled(const Consensus::Params &params,
+                   const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetIntArg("-ruthactivationtime", params.ruthActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 2cca9e2df4..a4767a8750 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -50,4 +50,8 @@ bool IsJoshuaEnabled(const Consensus::Params &params,
 bool IsJudgesEnabled(const Consensus::Params &params,
                      const CBlockIndex *pindexPrev);
 
+/** Check if 2024-12-21T09:20:00.000Z protocol upgrade has activated. */
+bool IsRuthEnabled(const Consensus::Params &params,
+                   const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index 95d31db7fd..a744e1e084 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -141,5 +141,9 @@ Consensus::CoinbaseAddresses AddressSets = {
         {
             "lotus_16PSJMaps9sQg7aBQgyY1RdHb2fZYdmWhQPbgus75",
         },
+    .ruth =
+        {
+            "lotus_16PSJPi88MtH34Ti3dZza4MFRF9XUVd3fKc6Ec3TV",
+        },
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 5263069ab6..1c65fd3161 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -25,6 +25,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> deuteronomy;
     std::vector<std::string> joshua;
     std::vector<std::string> judges;
+    std::vector<std::string> ruth;
 };
 
 enum BuriedDeployment : int16_t {
@@ -66,6 +67,9 @@ struct Params {
     /** Unit time used for MTP activation of 2024-12-21T09:20:00.000Z protocol
      * upgrade */
     int ruthActivationTime;
+    /** Unix time used for MTP activation of 2025-06-21T02:42:00.000Z protocol
+     * upgrade */
+    int firstSamuelActivationTime;
     /** Block height at which BIP16 becomes active */
     int BIP16Height;
     /** Block height and hash at which BIP34 becomes active */
diff --git a/src/init.cpp b/src/init.cpp
index 5749d35d57..52f4f5ad0f 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -472,7 +472,7 @@ void SetupServerArgs(NodeContext &node) {
         "-splash",
         "-uiplatform",
         // TODO remove after the May 2026 upgrade
-        "-obolenskyactivationtime", "-exodusactivationtime", "-leviticusactivationtime", "-numbersactivationtime", "-deuteronomyactivationtime", "-joshuaactivationtime", "-judgesactivationtime",
+        "-obolenskyactivationtime", "-exodusactivationtime", "-leviticusactivationtime", "-numbersactivationtime", "-deuteronomyactivationtime", "-joshuaactivationtime", "-judgesactivationtime", "-ruthactivationtime",
     };
 
     // Set all of the args and their help
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index fa7577686e..fb41b584a7 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -134,6 +134,8 @@ public:
         consensus.judgesActivationTime = 1718916600;
         // 2024-12-21T09:20:00.000Z protocol upgrade
         consensus.ruthActivationTime = 1734772800;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime = 1750473720;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -319,6 +321,10 @@ public:
         // 2024-12-21T09:20:00.000Z protocol upgrade
         consensus.ruthActivationTime =
             mainnetConsensus.ruthActivationTime - testnetActivationOffset;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime =
+            mainnetConsensus.firstSamuelActivationTime -
+            testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -473,6 +479,9 @@ public:
         consensus.judgesActivationTime = mainnetConsensus.judgesActivationTime;
         // 2024-12-21T09:20:00.000Z protocol upgrade
         consensus.ruthActivationTime = mainnetConsensus.ruthActivationTime;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime =
+            mainnetConsensus.firstSamuelActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index 9107522ef3..fa182daef4 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -60,6 +60,11 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
+    if (IsRuthEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.ruth,
+                                   pindexPrev, blockReward);
+    }
+
     if (IsJudgesEnabled(params, pindexPrev)) {
         return BuildOutputsCycling(params.coinbasePayoutAddresses.judges,
                                    pindexPrev, blockReward);
diff --git a/src/validation.cpp b/src/validation.cpp
index 79891cf828..8e9d3074ed 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -227,7 +227,7 @@ IsReplayProtectionEnabled(const Consensus::Params &params,
     }
 
     return pindexPrev->GetMedianTimePast() >=
-           activation_time.value_or(params.ruthActivationTime);
+           activation_time.value_or(params.firstSamuelActivationTime);
 }
 
 /**
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index 92523074fa..e453208a23 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -32,8 +32,9 @@ DEUTERONOMY_ACTIVATION_TIME = 2030000000
 JOSHUA_ACTIVATION_TIME = 2040000000
 JUDGES_ACTIVATION_TIME = 2050000000
 RUTH_ACTIVATION_TIME = 2060000000
+FIRST_SAMUEL_ACTIVATION_TIME = 2070000000
 
-REPLAYPROTECTION_ACTIVATION_TIME = RUTH_ACTIVATION_TIME
+REPLAYPROTECTION_ACTIVATION_TIME = FIRST_SAMUEL_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -126,6 +127,9 @@ JOSHUA_SCRIPTS = [
 JUDGES_SCRIPTS = [
     "76a9149208ecc785c92968481a92aee7024b77e54d27dc88ac",
 ]
+RUTH_SCRIPTS = [
+    "76a914d7756030f4292b70cb70dc4df231b14109c6806188ac",
+]
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -139,6 +143,7 @@ class MinerFundActivationTest(BitcoinTestFramework):
             f'-deuteronomyactivationtime={DEUTERONOMY_ACTIVATION_TIME}',
             f'-joshuaactivationtime={JOSHUA_ACTIVATION_TIME}',
             f'-judgesactivationtime={JUDGES_ACTIVATION_TIME}',
+            f'-ruthactivationtime={RUTH_ACTIVATION_TIME}',
             f'-replayprotectionactivationtime={REPLAYPROTECTION_ACTIVATION_TIME}',
         ]]
 
@@ -388,6 +393,37 @@ class MinerFundActivationTest(BitcoinTestFramework):
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Using the ruth addresses before upgrade fails
+        block = make_block_cb_post_numbers(RUTH_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+        node.setmocktime(RUTH_ACTIVATION_TIME)
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(JUDGES_SCRIPTS)
+            block.nTime = RUTH_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     RUTH_ACTIVATION_TIME)
+
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+            make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS),
+            make_block_cb_post_numbers(JOSHUA_SCRIPTS),
+            make_block_cb_post_numbers(JUDGES_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Using the Ruth scripts now works
+        for i in range(0, 12):
+            block = make_block_cb_post_numbers(RUTH_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
         # Check replay protection is not enabled yet
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
